### PR TITLE
Implement Decoder(generator, discriminator and loss)

### LINKED ISSUE #13: GAN style loss for training decoder in AutoEncoder
MS-STFT Discriminator needed

### PR DESCRIPTION
**Pull request by bwook00**
close #13

Plz review after pr merged #25 

While review, you don't need to read full codes.
Because I got this code from below repo

- https://github.com/jik876/hifi-gan/blob/master/models.py

If something wrong while training, I will fix it later.

### CODE REVIEW

File: songsuro/autoencoder/decoder/generator.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+						channels,
+						channels,
+						kernel_size,
+						1,
+						dilation=dilation[2],
+						padding=get_padding(kernel_size, dilation[2]),
+					)
+				),
+			]
+		)
```
Comment (vkehfdl1) [inline songsuro/autoencoder/decoder/generator.py:48]: The layer looks so repetitive, isn't it? use for-loop

File: songsuro/autoencoder/decoder/generator.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+					Conv1d(
+						channels,
+						channels,
+						kernel_size,
+						1,
+						dilation=1,
+						padding=get_padding(kernel_size, 1),
+					)
+				),
+			]
```
Comment (vkehfdl1) [inline songsuro/autoencoder/decoder/generator.py:83]: too repetitive also

File: songsuro/autoencoder/decoder/utils.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+def init_weights(m, mean=0.0, std=0.01):
+	classname = m.__class__.__name__
+	if classname.find("Conv") != -1:
+		m.weight.data.normal_(mean, std)
+
+
+def get_padding(kernel_size, dilation=1):
+	return int((kernel_size * dilation - dilation) / 2)
```
Comment (vkehfdl1) [inline songsuro/autoencoder/decoder/utils.py:8]: Is it better to move these to `utils` package?

**Review by vkehfdl1 — CHANGES_REQUESTED** [review]

File: songsuro/autoencoder/decoder/discriminator.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+from torch.nn.utils import weight_norm, spectral_norm
+
+from songsuro.autoencoder.decoder.utils import get_padding
+
+LRELU_SLOPE = 0.1
+
+
+class DiscriminatorP(torch.nn.Module):
+	def __init__(self, period, kernel_size=5, stride=3, use_spectral_norm=False):
+		super(DiscriminatorP, self).__init__()
```
Comment (minsing-jin) [inline songsuro/autoencoder/decoder/discriminator.py:14]: The current implementation uses super(ClassName, self).__init__(), which was necessary in Python 2. 

However, in Python 3, we can simplify this to super().__init__(), improving readability and maintaining the same functionality!

image

File: songsuro/autoencoder/decoder/discriminator.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+		x = self.conv_post(x)
+		fmap.append(x)
+		x = torch.flatten(x, 1, -1)
+
+		return x, fmap
+
+
+class MultiPeriodDiscriminator(torch.nn.Module):
+	def __init__(self):
+		super(MultiPeriodDiscriminator, self).__init__()
```
Comment (minsing-jin) [inline songsuro/autoencoder/decoder/discriminator.py:84]: we can simplify this to super().init(), improving readability and maintaining the same functionality!

File: songsuro/autoencoder/decoder/discriminator.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+			fmap_rs.append(fmap_r)
+			y_d_gs.append(y_d_g)
+			fmap_gs.append(fmap_g)
+
+		return y_d_rs, y_d_gs, fmap_rs, fmap_gs
+
+
+class DiscriminatorS(torch.nn.Module):
+	def __init__(self, use_spectral_norm=False):
+		super(DiscriminatorS, self).__init__()
```
Comment (minsing-jin) [inline songsuro/autoencoder/decoder/discriminator.py:113]: we can simplify this to super().init(), improving readability and maintaining the same functionality!

File: songsuro/autoencoder/decoder/discriminator.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+		x = self.conv_post(x)
+		fmap.append(x)
+		x = torch.flatten(x, 1, -1)
+
+		return x, fmap
+
+
+class MultiScaleDiscriminator(torch.nn.Module):
+	def __init__(self):
+		super(MultiScaleDiscriminator, self).__init__()
```
Comment (minsing-jin) [inline songsuro/autoencoder/decoder/discriminator.py:143]: we can simplify this to super().init(), improving readability and maintaining the same functionality!

File: songsuro/autoencoder/decoder/generator.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+from torch.nn import Conv1d, ConvTranspose1d
+from torch.nn.utils import weight_norm, remove_weight_norm
+from songsuro.autoencoder.decoder.utils import get_padding, init_weights
+
+LRELU_SLOPE = 0.1
+
+
+class ResBlock1(torch.nn.Module):
+	def __init__(self, h, channels, kernel_size=3, dilation=(1, 3, 5)):
+		super(ResBlock1, self).__init__()
```
Comment (minsing-jin) [inline songsuro/autoencoder/decoder/generator.py:13]: we can simplify this to super().init(), improving readability and maintaining the same functionality!

File: songsuro/autoencoder/decoder/generator.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+	def remove_weight_norm(self):
+		for layer in self.convs1:
+			remove_weight_norm(layer)
+		for layer in self.convs2:
+			remove_weight_norm(layer)
+
+
+class ResBlock2(torch.nn.Module):
+	def __init__(self, h, channels, kernel_size=3, dilation=(1, 3)):
+		super(ResBlock2, self).__init__()
```
Comment (minsing-jin) [inline songsuro/autoencoder/decoder/generator.py:105]: we can simplify this to super().init(), improving readability and maintaining the same functionality!

File: songsuro/autoencoder/decoder/generator.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+		return x
+
+	def remove_weight_norm(self):
+		for layer in self.convs:
+			remove_weight_norm(layer)
+
+
+class Generator(torch.nn.Module):
+	def __init__(self, h):
+		super(Generator, self).__init__()
```
Comment (minsing-jin) [inline songsuro/autoencoder/decoder/generator.py:147]: we can simplify this to super().init(), improving readability and maintaining the same functionality!

**Review by minsing-jin — CHANGES_REQUESTED** [review]
With a few modifiy, I think it'll be LGTM

image

**Comment by minsing-jin** [comment]
How about leaving the source of the code as a comment?

**Comment by bwook00** [comment]
- Add source and license of the code
- Apply HiddenSinger's Decoder loss logic
- Apply other resolve point

**Review by vkehfdl1 — APPROVED** [review]
Awesome kim

**Review by minsing-jin — APPROVED** [review]
LGTM

image
